### PR TITLE
Optimized Session Suggestion with Conflict Resolution Fallback

### PR DESCRIPTION
**Pull request by Fis-ayo**
## Description

This PR implements an optimized session suggestion that first tries to find direct mutual free time between mentee and mentor. If that fails, it intelligently resolves conflicts by suggesting to reschedule an existing session to make a free time with the new mentor (connection).

### Context
**Refactored the two suggestion logic into two clear phases**:
- Direct Suggestion Phase:
   - Uses `getAvailability`, `getFreeSlots`, `timeOverlaps`, and `oneHourIntervals` to find a clean 1-hour mutual time block.
   - Return immediately at most `5` suggestion if found.
- Conflict Resolution Phase
   - Constructs a flattened, sorted `allSession` of all connected mentor sessions.
   - Iterates through valid free slots, using `filterOverlaps()` to check if a single session is blocking the slot.
   - If blocked, attempts to reschedule that sessions into another viable 1-hour time window (using helper functions).
   - Uses memoization (`mentorCache`) to minimize redundant Prisma DB calls.
**Refactoring & Performance**:
- Initially had deeply nested loops to resolve conflicts(`conflict x users x sessions x options`). Eliminated them.
- Replaced magic numbers with name constants in `config/constants.js`.
- Preserved original utility functions to maintain consistency with existing logic.

### Future PRs
- **Unit tests**:
   - Handle edge cases for direct session suggestion algorithm.
   - Ensure correct conflict detection and safe rescheduling when resolving conflicts.

## Milestones
This PR completes majority of the backend algorithm to help users calculate free times and give suggestions as well as resolving conflicts when there are no mutual free slots between a mentee and a mentor.

## Resources
[Memoization](https://www.geeksforgeeks.org/dsa/what-is-memoization-a-complete-tutorial/)

## Test Plan
- Future unit test would include prove that works
- As of now, I used two test cases shown in the **Insomnia result**
### Test Artifacts
- Mentor with no conflicting mutual free time.
- Mentor with conflicting session

Screenshot 2025-07-19 at 10 37 38 PM

Screenshot 2025-07-19 at 10 38 09 PM

### CODE REVIEW

File: backend/controllers/sessionController.js [hunk 1]
```diff
@@ -1,38 +1,44 @@
...
+const { getSessions, getAvailability, getFreeSlots } = require('../utils/scheduleHelpers');
+const resolveConflict = require('../utils/resolveConflict');
+const { TOP_NUMBER } = require('../config/constants');
 
 const suggestSession = async (req, res) => {
     const menteeId = req.session.userId;
     const mentorId = parseInt(req.params.mentorId);
+
+    if (Number.isNaN(mentorId)) {
+        return res.status(400).json({ error: "Query is not a number" });
```
Comment (chandra-tacc) [inline backend/controllers/sessionController.js:12]: Instead of this "Query is not a number", can you change to "MentorId is not a number"?

File: backend/utils/schedulingUtils.js [hunk 2]
```diff
@@ -64,14 +27,14 @@ const timeOverlaps = (menteeSlots, mentorSlots) => {
         a = menteeSlots[i];
         b = mentorSlots[j];
```
Comment (chandra-tacc) [inline backend/utils/schedulingUtils.js:28]: This is a bug. a and b are used without let or const, they will become global variables

File: backend/utils/schedulingUtils.js [hunk 3]
```diff
@@ -81,34 +44,82 @@ const timeOverlaps = (menteeSlots, mentorSlots) => {
...
+ */
+
+const timeConversion = (day, time) => {
+    const today = new Date();
+    const todayDay = today.getDay();
+
+    const targetDay = DayMap[day];
+
+    let daysAhead = (targetDay - todayDay + DAYS_IN_WEEK) % DAYS_IN_WEEK;
+    if(daysAhead === 0) daysAhead = DAYS_IN_WEEK;
```
Comment (chandra-tacc) [inline backend/utils/schedulingUtils.js:85]: if daysAhead is zero, you are always setting this to next week. Is this intended?

File: backend/utils/schedulingUtils.js [hunk 4]
```diff
@@ -64,14 +27,14 @@ const timeOverlaps = (menteeSlots, mentorSlots) => {
         a = menteeSlots[i];
         b = mentorSlots[j];
 
-        start = timeToMinutes(a[0]) > timeToMinutes(b[0]) ? a[0] : b[0];
-        end = timeToMinutes(a[1]) < timeToMinutes(b[1]) ? a[1] : b[1];
+        start = a[0] > b[0] ? a[0] : b[0];
+        end = a[1] < b[1] ? a[1] : b[1];
```
Comment (chandra-tacc) [inline backend/utils/schedulingUtils.js:31]: same with start and end. Please use `let`

File: backend/utils/scheduleHelpers.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * Function to get all exisiting sessions of a user
+ * 
+ * @param {Array<{Object}>} mentorshipList - List of user's connections and their sessions
+ * @returns {Array<[Number, Number]>} - Array of sorted user sessions
+ * 
+ */
+
+const getSessions = (mentorshipList) => {
+    const sessions = [];
+    mentorshipList.map(record => {
```
Comment (chandra-tacc) [inline backend/utils/scheduleHelpers.js:13]: .map is for transforming arrays. here you populating sessions, change to .forEach instead of map - in line 13 and 14.

File: backend/utils/schedulingUtils.js (hunk 3 above)
Comment (Fis-ayo) [reply backend/utils/schedulingUtils.js:85]: Yes. Say today is Monday, it'll get results from the monday.

File: backend/utils/schedulingUtils.js (hunk 4 above)
Comment (Fis-ayo) [reply backend/utils/schedulingUtils.js:31]: Start and end have already being declared at the top.

**Review by chandra-tacc — APPROVED** [review]
